11 - Custom Video Player/scripts.js

```diff
@@ -0,0 +1,103 @@
+        this.toggle.textContent = evt.target.paused ? '►' : '❚ ❚';
+    };
+    
+    skipSeconds = (evt) => {
+        this.video.currentTime += parseFloat(evt.target.dataset.skip);
+    };
+
+    handleRanges = (evt) => {
+        const { name, value } = evt.target;
+        this.video[name] = value;
+    };
+
+    handleTime = (evt) => {
+        const progressPercent = (this.video.currentTime / this.video.duration) * 100;
+        this.progressBar.style.flexBasis = `${progressPercent}%`;
+    };
+
+    dragProgress = (evt) => {
+        const scrubTime = (evt.offsetX / this.progress.offsetWidth) * this.video.duration;
+        this.video.currentTime = scrubTime;
+      };
+
+    toggleMouseDown = () => {
+        this.mouseDown = !this.mouseDown;
+    };
+
+    handleFullScreen = () => {
+        document.fullscreenElement = 
+            document.fullscreenElement || 
+            document.mozFullscreenElement || 
+            document.msFullscreenElement || 
+            document.webkitFullscreenDocument;
+
+        document.exitFullscreen = 
+            document.exitFullscreen || 
+            document.mozExitFullscreen || 
+            document.msExitFullscreen || 
+            document.webkitExitFullscreen;
+
+        this.video.requestFullscreen = 
+            this.video.requestFullscreen || 
+            this.video.mozRequestFullscreen || 
+            this.video.msRequestFullscreen || 
+            this.video.webkitRequestFullscreen;
+
+        
+        if (!document.fullscreenElement) {
+            this.video.requestFullscreen()
+                .catch(err => {
+                    alert(`Error attempting to enable full-screen mode: ${err.message} (${err.name})`);
+                });
+        } else {
+          if (document.exitFullscreen) {
+            document.exitFullscreen();
+          }
+        }
+        
+    };
+};
+
+//// class init
+const newVideoPlayer = new CustomVideoPlayer('.player');
+
+//// listeners
+// play-pause handlers
+newVideoPlayer.video.addEventListener('click', newVideoPlayer.playPause);
+newVideoPlayer.toggle.addEventListener('click', newVideoPlayer.playPause);
+newVideoPlayer.video.addEventListener('play', newVideoPlayer.toggleIcon);
+newVideoPlayer.video.addEventListener('pause', newVideoPlayer.toggleIcon);
+
+// volume and speed handlers 
+newVideoPlayer.ranges.forEach(range => range.addEventListener('change', newVideoPlayer.handleRanges));
+newVideoPlayer.ranges.forEach(range => range.addEventListener('mousemove', newVideoPlayer.handleRanges));
+
+// skip, progress and time handlers
+newVideoPlayer.skipButtons.forEach(skipButton => skipButton.addEventListener('click', newVideoPlayer.skipSeconds));
+newVideoPlayer.video.addEventListener('timeupdate', newVideoPlayer.handleTime);
+newVideoPlayer.progress.addEventListener('click', newVideoPlayer.dragProgress);
+newVideoPlayer.progress.addEventListener('mousemove', () => newVideoPlayer.mouseDown && newVideoPlayer.dragProgress);
```

Comment:
here another example of logic that semantically belongs into the class. or do you feel it is necessary to do this outside?

Reply:
same as above. I'd like to trigger everything from the `player` but I did not find a better solution.

Reply:
let me know if you keep having problems even when you register the events inside the class.

Reply:
I moved all the event listeners inside of the constructor :) 
I actually didn't know you can do that... but today I woke up and I thought I give it a try and it works.

Too extreme? I think this exercise fits perfectly the encapsulation principle because the `CustomVideoPlayer` can be seen as a widget and reused anywhere in the code with the same functionalities.

If the developer needs small variations to this player, they could be implemented with a sub-class or simply changing the main class based on what is needed.

Reply:
I think this class is much more encapsulated than before. nice work. maybe for the next time split constructor into multiple methods and see how you like it better. Example:

```
construction(rootElm) {
  this.playButton = rootElm.querySelector('[data-role="play"]');
  // more elements

  setupListeners();
  doOtherStuff();
```